File: plugins/typescript/nginx-lint-plugin/src/include-context.ts
/**
 * Shared include-context helpers for Config implementations.
 *
 * Provides the include-context portion of the Config interface so that
 * config-builder.ts can reuse a single, tested implementation.
 */

export interface IncludeContextMethods {
  includeContext(): string[];
  isIncludedFrom(context: string): boolean;
  isIncludedFromHttp(): boolean;
  isIncludedFromHttpServer(): boolean;
  isIncludedFromHttpLocation(): boolean;
  isIncludedFromStream(): boolean;
  immediateParentContext(): string | undefined;
}

export function makeIncludeContextMethods(inclCtx: string[]): IncludeContextMethods {
  return {
    includeContext() { return inclCtx; },
    isIncludedFrom(context: string) { return inclCtx.includes(context); },
    isIncludedFromHttp() { return inclCtx.includes("http"); },
    isIncludedFromHttpServer() {
      const httpIndex = inclCtx.indexOf("http");
      const serverIndex = inclCtx.indexOf("server");
      return httpIndex !== -1 && serverIndex !== -1 && httpIndex < serverIndex;
    },
    isIncludedFromHttpLocation() {
      const httpIndex = inclCtx.indexOf("http");
      const locationIndex = inclCtx.indexOf("location");
      return httpIndex !== -1 && locationIndex !== -1 && httpIndex < locationIndex;
    },
    isIncludedFromStream() { return inclCtx.includes("stream"); },
    immediateParentContext() { return inclCtx.length > 0 ? inclCtx[inclCtx.length - 1] : undefined; },
  };
}

File: plugins/python/nginx-lint-plugin/python/nginx_lint_plugin/include_context.py
"""Shared include-context helpers for Config implementations.

Port of the TypeScript SDK's include-context.ts: provides the
include-context portion of the Config interface so config_builder can
reuse a single implementation.
"""

from typing import List, Optional


class IncludeContextMixin:
    """Mixin implementing the include-context methods of the WIT config-api.

    Subclasses must set ``self._include_context`` (a list of parent block
    names) before use.
    """

    _include_context: List[str]

    def include_context(self) -> List[str]:
        return self._include_context

    def is_included_from(self, context: str) -> bool:
        return context in self._include_context

    def is_included_from_http(self) -> bool:
        return "http" in self._include_context

    def is_included_from_http_server(self) -> bool:
        ctx = self._include_context
        return "http" in ctx and "server" in ctx and ctx.index("http") < ctx.index("server")

    def is_included_from_http_location(self) -> bool:
        ctx = self._include_context
        return "http" in ctx and "location" in ctx and ctx.index("http") < ctx.index("location")

    def is_included_from_stream(self) -> bool:
        return "stream" in self._include_context

    def immediate_parent_context(self) -> Optional[str]:
        return self._include_context[-1] if self._include_context else None
